File: Python/TraceBus.py
from typing import List


class TraceBus:
    def __init__(self):
        self.listeners: List = []

    def add_listener(self, listener) -> None:
        self.listeners.append(listener)

    def publish(self, event) -> None:
        for l in self.listeners:
            if hasattr(l, 'on_event'):
                l.on_event(event)
            elif hasattr(l, 'onEvent'):
                l.onEvent(event)

File: Java/TraceBus.java
package mu.cse.rag.clean;

import java.util.ArrayList;
import java.util.List;

public class TraceBus {

    private final List<TraceListener> listeners = new ArrayList<>();

    public void addListener(TraceListener listener) {
        listeners.add(listener);
    }

    public void publish(TraceEvent event) {
        for (TraceListener l : listeners) {
            l.onEvent(event);
        }
    }
}
